File: src/api/private/instance-parameters.ts
import { RequestFN } from "~/core/request-function";
import type { SessionHandle, InstanceParameters } from "~/models";
import forge from "node-forge";
import { decodeInstanceParameters } from "~/decoders/instance-parameters";

export const instanceParameters = async (session: SessionHandle, navigatorIdentifier: string | null = null): Promise<InstanceParameters> => {
  const rsaKey = forge.pki.rsa.setPublicKey(
    new forge.jsbn.BigInteger(session.information.rsaModulus, 16),
    new forge.jsbn.BigInteger(session.information.rsaExponent, 16)
  );

  const aesIV = session.information.aesIV;
  let uuid: string;

  if (session.information.rsaFromConstants) {
    uuid = forge.util.encode64(session.information.http ? rsaKey.encrypt(aesIV) : aesIV, 64);
  }
  else {
    uuid = forge.util.encode64(rsaKey.encrypt(aesIV));
  }

  const request = new RequestFN(session, "FonctionParametres", {
    donnees: {
      identifiantNav: navigatorIdentifier,
      Uuid: uuid
    }
  });

  const response = await request.send();
  return decodeInstanceParameters(response.data.donnees);
};

File: library/src/jvmMain/kotlin/ink/literate/pawnote/api/private/instance-parameters.kt
package ink.literate.pawnote.api.private

import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.decoders.decodeInstanceParameters
import ink.literate.pawnote.models.InstanceParameters
import ink.literate.pawnote.models.SessionInformation
import java.math.BigInteger
import java.security.KeyFactory
import java.security.spec.RSAPublicKeySpec
import javax.crypto.Cipher
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

@OptIn(ExperimentalEncodingApi::class, ExperimentalStdlibApi::class)
actual suspend fun instanceParameters(
    sessionInfo: SessionInformation,
    navigatorIdentifier: String?
): InstanceParameters {
  val modulus = BigInteger(sessionInfo.rsaModulus, 16)
  val exponent = BigInteger(sessionInfo.rsaExponent, 16)

  val pubKeySpec = RSAPublicKeySpec(modulus, exponent)
  val pubKey = KeyFactory.getInstance("RSA").generatePublic(pubKeySpec)

  val aesIV = sessionInfo.aesIV
  val uuid: String

  val cipher = Cipher.getInstance("RSA")
  cipher.init(Cipher.ENCRYPT_MODE, pubKey)

  uuid =
      if (sessionInfo.rsaFromConstants)
          Base64.encode(
              if (sessionInfo.http) cipher.doFinal(aesIV.hexToByteArray())
              else aesIV.hexToByteArray())
      else Base64.encode(cipher.doFinal(aesIV.hexToByteArray()))

  val requestData = buildJsonObject {
    putJsonObject("donnees") {
      put("identifiantNav", navigatorIdentifier)
      put("Uuid", Json.parseToJsonElement(uuid))
    }
  }

  val request = RequestFN(sessionInfo, "FonctionParametres", Json.encodeToString(requestData))
  val response = request.send()

  return decodeInstanceParameters(
      Json.parseToJsonElement(response.data).jsonObject["donnees"]!!.jsonObject)
}
